File: lambda/recupero/python/recupero.py
import psycopg2
from datetime import datetime
import os

class ReactionTime:
    def __init__(self, id: int, time: int, tms_insert: datetime, user_id: int, name: str):
        self.id = id
        self.time = time
        self.tms_insert = tms_insert
        self.user_id = user_id
        self.name = name

    def to_dict(self):
        return {
            "id": self.id,
            "time": self.time,
            "tms_insert": self.tms_insert.strftime("%Y-%m-%d %H:%M:%S"),
            "user_id": self.user_id,
            "name": self.name
        }

def lambda_handler(event, context):
    response = []
    with psycopg2.connect(dsn=os.getenv("DATABASE_DSN")) as conn:
        reaction_times = get_all_reaction_times(conn)

        for i, time in enumerate(reaction_times):
            response.append(ReactionTime(time[0], time[1], time[2], time[3], time[4]).to_dict())   #[i] = ReactionTime(time[0], time[1], time[2], time[3], time[4]).to_dict()


    return response


def get_all_reaction_times(db_connection):
    try:
        cursor = db_connection.cursor()
        cursor.execute("SELECT reaction_times.id, reaction_times.time, reaction_times.tms_insert, reaction_times.user_id, users.name FROM reaction_times JOIN users ON reaction_times.user_id = users.id WHERE (reaction_times.user_id, reaction_times.time) IN (SELECT user_id, MIN(time) FROM reaction_times GROUP BY user_id) ORDER BY reaction_times.time ASC;")
        records = cursor.fetchall()
        return records
    except Exception as e:
        print("Errore durante il recupero dei dati:", e)
        return []

File: lambda/recupero/java/src/main/java/recupero/Recupero.java
package recupero;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestStreamHandler;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.postgresql.Driver;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Add Value S.p.A. by nicholas.mantovani
 * @version apr  11, 2024
 */
public class Recupero implements RequestStreamHandler {
    private ObjectMapper objectMapper = new ObjectMapper();

    private static final String GET_REACTION_TIME = "SELECT reaction_times.id, reaction_times.time, reaction_times.tms_insert, reaction_times.user_id, users.name FROM reaction_times JOIN users ON reaction_times.user_id = users.id WHERE (reaction_times.user_id, reaction_times.time) IN (SELECT user_id, MIN(time) FROM reaction_times GROUP BY user_id) ORDER BY reaction_times.time ASC;";

    @Override
    public void handleRequest(InputStream inputStream, OutputStream outputStream, Context context) throws IOException {

        System.out.println("Lambda started");
        Connection connection = null;
        PreparedStatement preparedStatement = null;
        ResultSet resultSet = null;

        List<ReactionTime> reactionTimes = new ArrayList<>();

        try {
            connection = getConnection();
            preparedStatement = connection.prepareStatement(GET_REACTION_TIME);

            resultSet = preparedStatement.executeQuery();

            while (resultSet.next()) {
                ReactionTime reactionTime = new ReactionTime();
                reactionTime.setId(resultSet.getLong("id"));
                reactionTime.setTime(resultSet.getLong("time"));
                reactionTime.setTmsInsert(resultSet.getDate("tms_insert"));
                reactionTime.setUserId(resultSet.getLong("user_id"));
                reactionTime.setName(resultSet.getString("name"));
                reactionTimes.add(reactionTime);
            }

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            close(connection, preparedStatement, resultSet);
        }

        System.out.println("Lambda ended, sending response");

        objectMapper.writeValue(outputStream, reactionTimes);
    }

    private Connection getConnection() throws SQLException {

        String databaseJdbcUrl = System.getenv("DATABASE_JDBC_URL");

        DriverManager.registerDriver(new Driver());

        return DriverManager.getConnection(databaseJdbcUrl);
    }

    public static void close(Connection connection,
                             PreparedStatement prepareStatement, ResultSet resultSet) {
        if (prepareStatement != null)
            try {
                prepareStatement.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        if (connection != null)
            try {
                connection.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        if (resultSet != null)
            try {
                resultSet.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
    }
}

